[ADR] Against the use of Lightweight Transaction in Cassandra code base (#271)
apache/james-project: src/adr/0044-against-the-use-of-cassandra-lightweight-transactions.md

that we might want to challenge:

 - `users` we rely on LWT for throwing "AlreadyExist" exceptions. LWT are likely unnecessary as the webadmin 
presentation layer is offering an idempotent API (and silents the AlreadyExist exceptions). Only the CLI
(soon to be deprecated for Guice products) makes this distinction. Discussions have started on the topic and a proof of
concept is available.
 - `domains` we rely on LWT for throwing "AlreadyExist" exceptions. LWT are likely unnecessary as the webadmin 
presentation layer is offering an idempotent API (and silents the AlreadyExist exceptions). Only the CLI
(soon to be deprecated for Guice products) makes this distinction. Discussions have started on the topic and a proof of
concept is available.
 - `mailboxes` relies on LWT to enforce name unicity. We hit the same pitfalls than for ACLs as this is a very often
 read table (however mailboxes of a given user being grouped together, primary key read are more limited hence this is
 less critical). Similar results could be expected. Discussions on this topic have not been started yet. Further

>  impact studies on performance needs to be conducted.

Reviewer: ```suggestion
 impact studies on performance need to be conducted.
```

flags we could be relying on Cassandra to solve data race issues for us. Note also that IMAP CONDSTORE extension is not 
implemented, and might be a non-viable option performance-wise. We might choose to favor performance other 
transactionality on this topic. Discussions on this topic have not started yet.

LWT are required for `eventSourcing`. As event sourcing usage is limited to low-usage use cases, the performance
degradations are not an issue.

LWT usage is required to generate `UIDs`. As append message operations tend to be limited compared to
message update operations, this is likely less critical. UID generation could be handled via alternative systems,
past implementations have been conducted on ZooKeeper.

If not implementing IMAP CONDSTORE, generation of IMAP `MODSEQ` likely no longer makes sense. As such the fate of
`MODSEQ` is linked to decisions on the `message` topic.

Similarly, LWT are used to try to keep the count of emails in MailRepository synchronize. Such a usage is non-performance
critical for a MDA (Mail Delivery Agent) use case but might have a bigger impact for MTA (Mail Transfer Agent). No

> discussion not work have been started on the topic.

Reviewer: ```suggestion
discussion or work have been started on the topic.
```

> Other usage of LWT includes Sieve script management, initialization of the RabbitMQMailQueue browse start and other

Reviewer: ```suggestion
Other usages of LWT include Sieve script management, initialization of the RabbitMQMailQueue browse start and other
```
